intro to eth pt. 2 + 3

wolovim opened the pull request:
## what does it do?

- Ports over intro to eth, parts 2 & 3. includes final quizzes, but no intermediary ones yet.
- Originally I broke them out into another "track", but v1 doesn't support that architecture well. Abandoned that idea for now and kept the lessons within the fundamentals directory. Punting until v2 to figure out the best home for them.

wolovim committed:
feat: introduce intro to ethereum track

wolovim committed:
fix: format eth intro pt 2

wolovim committed:
fix: remove eth intro pt 1 from fundamentals

wolovim committed:
feat: introduce eth intro pt 3

wolovim commented:
@Markkos89 @okhaimie-dev @peteruche21 at your convenience, let me know your opinion on the pattern ive used to introduce another track and if youd recommend something different for v1. the camel case URL definitely needs fixing either way (e.g., `/lessons/ethIntro/2`). warning: my JS knowledge is stale/outdated 😅

wolovim committed:
chore: rename eth-intro route

wolovim committed:
feat: start part 2 quiz

wolovim committed:
feat: finish pt 2 quiz + undo new routes for v1

wolovim committed:
feat: pt 3 quiz

wolovim committed:
chore: add pt 3 formatting + rename lesson files

elPiablo commented:
> ## what does it do?
> * Ports over intro to eth, parts 2 & 3. includes final quizzes, but no intermediary ones yet.
> * Originally I broke them out into another "track", but v1 doesn't support that architecture well. Abandoned that idea for now and kept the lessons within the fundamentals directory. Punting until v2 to figure out the best home for them.

I would love to see them in a track of their own. Shining!

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-2.mdx:
```diff
@@ -0,0 +1,283 @@
+- What makes Ethereum decentralized?
+- What is ether and why is it a necessary component of the network?
+
+In this post, we’ll build on these concepts and explore some of the implications
```
can you change 'post' to lesson or tutorial?

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-2.mdx:
```diff
@@ -0,0 +1,283 @@
+
+## Web2 vs. Web3
+
+It’s hard not to collect accounts in today’s web. You have one for each social media app, news site, delivery service, retailer, and airline — to name a few. Each of these accounts lives on the company's servers, which makes you subject to their terms and conditions, privacy policy, and security practices. Your account can be frozen, deleted, censored, or altered at the discretion of the host com ...
```
Could you change 'airline' to something like 'public transport pass' so we're promoting environmentally friendly practices?

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-2.mdx:
```diff
@@ -0,0 +1,283 @@
+## Creating an account
+
+Same drill as last time: the concepts will be demonstrated in an IPython shell. If you’re not a Python developer, no problem. Just follow along conceptually.
+
```
It would be nice to continue along the theme of using a d_d_academy directory for learners' projects - also to promote good practices. It would mean a slight adjustment on the first installment of Eth Intro series

wolovim replied:
we can revisit this, but the code is transient for the sake of conveying the concepts. there is no project to speak of, so no files to create or code to store. for now, i'll suggest we punt on this, in the interest of shipping a 'good enough' first iteration.

wolovim committed:
chore: pr feedback + external link removal

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-2.mdx:
```diff
@@ -0,0 +1,279 @@
+### Environment setup
+
+Three steps to set the stage:
+
```
Marc, the word `eth-tester` here is confusing. Do you mean `web3-tester`?

The sentence would also be much easier to understand,  by omitting the brackets (because it lowers the priority for the reader, or at least to me) and swap the clauses round, i.e. 'In case you didn't already in Part 1, install web3py, web3-tester and iPython:' (tbh, I was here for a good 30 mins back and forth trying to figure out what was going on)

elPiablo replied:
I finally found what eth-tester was referring to...

elPiablo replied:
What I was saying here, Marc, is that I was completely lost here for a long time. It was very frustrating. Not good for learners. Let's come back to this at a later date - this situation is something we should document as it is fundamental to the review process.

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-2.mdx:
````diff
@@ -0,0 +1,279 @@
+Let's create that account:
+
+```python
+In [2]: w3 = Web3() # No provider necessary for this demo
````
I believe we need to make it explicit to paste in the `# No provider necessary for this demo` other wise we 'do' end up with an output. 

I got the following, and went on to correct it: 
`In [1]: from web3 import Web3

In [2]: Web3()
Out[2]: <web3.main.Web3 at 0x7fd1d1d24c10>

In [3]: acct = w3.eth.account.create()
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[3], line 1
----> 1 acct = w3.eth.account.create()

NameError: name 'w3' is not defined

In [4]: acct.address
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[4], line 1
----> 1 acct.address

NameError: name 'acct' is not defined

In [5]: w3 = Web3() # No provider necessary for this demo

In [6]: acct = w3.eth.account.create()

In [7]: acct.address
Out[7]: '0xD8Bf88bd36D253167c5f129faAa1D09De91671f2'

In [8]: acct.key
Out[8]: HexBytes('0x94c789f4003025f345f3eecd519b84425eb654a1b5f731dbc465c22e38a98dd5')

In [9]: 
``

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-2.mdx:
```diff
@@ -0,0 +1,279 @@
+
+### Transferring ether
+
+Recall that `EthereumTesterProvider` enables a test environment seeded with accounts and fake ether. Let's start by viewing some test accounts and an account balance:
```
Do you want to make this a question by replacing with a question mark at the end instead of full stop? It will personalise it for the learner a lot :)

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-2.mdx:
````diff
@@ -0,0 +1,279 @@
+```
+
+Let's break that down. Step 1 defines a Python dictionary with the required transaction fields. We briefly learned about gas (transaction fees) in Part 1, but nonce may be new territory. In Ethereum, a **nonce** is simply the transaction count of the account. The Ethereum protocol keeps track of this value to prevent double-spending.
+
````
Would you mind using this as a very small learning moment to educate non-Python-heads what a dictionary is in Python, e.g. in brackets after '...required transaction fields (A dictionary, or `dict` is a Python type similar to an object in other programming languages).' And then consolidate it three paragraphs further down with '...a from value is missing from the `tx` dict'. Just an idea.

elPiablo replied:
Lovely job!!

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-2.mdx:
````diff
@@ -0,0 +1,279 @@
+```
+
+Besides requiring more gas, the only other distinction in a contract deployment transaction is the absence of a `to` value. The rest of the process is identical to a standard transfer of ether.
+
````
Is there no `to` value because the smart contract's deployment creates the address it's going to go `to` live in?

elPiablo replied:
Shmoooooothly done!

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-2.mdx:
```diff
@@ -0,0 +1,279 @@
+Transactions are the only way to affect the state of the blockchain, but they aren't the only way an account can be utilized. Simply proving ownership of a particular account can be useful in its own right.
+
+As an example, OpenSea, an Ethereum marketplace, will allow you to bid on items for sale by signing a message with your account. Only when an auction expires or the seller accepts your offer is an actual transaction placed. Similarly, the app uses signed messages as a form of authentication before showing you some account details.
+
```
does the 'app' refer to OpenSea?

elPiablo replied:
Nice!

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-2.mdx:
```diff
@@ -0,0 +1,279 @@
+
+<br />
+<QuizStatusChecker quiz="quiz-eth-intro-2" />
+
```
Can you check question 6 of the final quiz, and the fourth answer option, which reads: 'A password for your account' - it reads a bit funny, given it doesn't start with a verb like the other answers' options?

elPiablo replied:
I think it's a really nice quiz.

I was a bit worried when I hit question 9 - 'A transaction to deploy a new contract omits which value?'- I read 'emits' as opposed to 'omits', and I thought that it was a very 'specific information' question, and therefore panicked trying to think of the answer :D.

wolovim replied:
good catch; i think the question six outlier answer was a copy/paste mistake. deleted it.

elPiablo replied:
The reason I initially got confused is because of 'emit'ting events:
https://ethereum.org/en/developers/tutorials/logging-events-smart-contracts/
I am a believer in tripping up learners, but actually crafting those situations intentionally and thoughtfully. But don't think we should be leaving anyone confused when we are aware of that confusion. Initial confusion is still confusion. 
A synonym for 'omit' could have been found. e.g. 'leave out', 'exclude', etc.  But let's come back to this sort of thing later in the form of a process 😄

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-2.mdx:
```diff
@@ -0,0 +1,279 @@
+>
+<LessonHeader title="A Developer's Guide to Ethereum, Pt. 2" />
+
+Welcome back! In the first installment of this series we covered a lot of ground
```
We could do with a header here for this first paragraph. Other than that, the use of ## headers and ### sub-headers are really meaningfully structured and therefore helpful for @uma-l and @markkos and dev team.

elPiablo commented:
I was surprised that I only got as far as the end of Part 2 tonight, @wolovim 😢

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-2.mdx:
````diff
@@ -0,0 +1,279 @@
+```
+
+This transaction will execute immediately, but some important details are hidden from view. web3.py is smart enough to know that `EthereumTesterProvider` is managing `acct_one` and that we're using a test environment. For our convenience, `acct_one` is "unlocked," meaning transactions from the account are approved (signed) by default.
+
````
Change `"unlocked,"`` `to "unlocked",``

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-2.mdx:
```diff
@@ -0,0 +1,279 @@
+      'from': acct_one,
+      'to': acct_two.address,
+      'value': Web3.toWei(1, 'ether')
+   })
```
Can you change this from `toWei` to `to_wei`?

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-3.mdx:
```diff
@@ -0,0 +1,209 @@
+
+## Smart contracts, an introduction
+
+The Ethereum blockchain has a great deal of value flowing through it. In previous articles, we discussed the flow of **ether** from one user to another via transactions. However, the network is capable of more sophisticated interactions, and those use cases are enabled by smart contracts.
```
I would define this with 'the', so we're pointing to part 1 and 2 that you just referred to - `In 'the' previous lessons, ...`. (you chose 'lesson', as opposed to 'post' in part 2, if I remember correctly :))

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-3.mdx:
```diff
@@ -0,0 +1,209 @@
+
+## Smart contracts, an introduction
+
+The Ethereum blockchain has a great deal of value flowing through it. In previous articles, we discussed the flow of **ether** from one user to another via transactions. However, the network is capable of more sophisticated interactions, and those use cases are enabled by smart contracts.
```
'The Ethereum blockchain has a great deal of value flowing through it, financial and otherwise' (might be a nice addidtion?)

elPiablo replied:
I would define this with 'the', so we're pointing to part 1 and 2 that you just referred to - `In 'the' previous lessons, ...`. (you chose 'lesson', as opposed to 'post' in part 2, if I remember correctly :))

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-3.mdx:
```diff
@@ -0,0 +1,209 @@
+
+In this paradigm, that's a requirement. Users (or other contracts) may utilize your contract to move real value around. They need to be able to trust that your contract does what you say it does, and in order to do that, they need to be able to read it for themselves.
+
+In reality, most users aren't reading the source code of each smart contract they interact with, but most wisely won't touch a contract if that source code isn't verified (e.g., on sourcify or Etherscan) and vetted (e.g., audited) by industry veterans.
```
Capital `S` in `Sourcify`

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-3.mdx:
````diff
@@ -0,0 +1,209 @@
+
+```solidity
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
````
Do you wanna just update the pragma to 0.8.20, since OpenZepp's version 5 is using 0.8.20 and upwards? Just an idea to embed things in first-timer's minds.

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-3.mdx:
```diff
@@ -0,0 +1,209 @@
+
+Each language is worth a look, but below is a "Hello, World"-style example written in Solidity, the most established of the languages. This example contract is titled `Billboard`, stores a single message, and contains one function to update that message. As written, anyone has the unrestricted ability to update that message.
+
+You can imagine that a website might display whatever message is stored and provide an input to type in a new message, replacing the current one. The combination of a smart contract and its user interface is what's referred to as a decentralized application, or "dapp" for short.
```
In order to make the previous paragraph flow more easily to this one, I would change this to `Imagine a website displaying whatever message is stored and providing an input ...` - otherwise the previous paragraph's closing sentence is powerful enought to do just that - close it of. And the ensuing paragraph reads like it could be a completely different subject. And you could even join the two paragraphs and make one since the reader is in 'imagine' mode. I think that would make a better all-round connection of the text to the ensuing  code block

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-3.mdx:
```diff
@@ -0,0 +1,209 @@
+
+If you're familiar with object-oriented programming, a contract will look an awful lot like the concept of a class. In effect, when a contract is deployed, a single instance becomes available for all the world to use — analogous to a "singleton" class. 
+
+For all users, a deployed contract has a particular state at any given block. In other words, the Billboard's `message` is always the same for everyone. A contract's state can keep track of all manner of things; within a token contract, for example, the state might include who owns how many of which assets.
```
For some conolidation, maybe add `on the chain` i.e. `... a particular state at any given block on the chain.`?

elPiablo replied:
` In other words, the Billboard's message will always read the same for everyone, until someone updates it. ` Would something like that be correct to assume?

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-3.mdx:
```diff
@@ -0,0 +1,209 @@
+
+In a Solidity contract, the `constructor` method is executed only once, when the contract is first deployed. Continuing the class analogy, the `constructor` might remind you of the `__init__` method within a Python class or similar initialization methods in other languages. So, whoever deploys this contract gets to determine the starting billboard message.
+
+You may have noticed the JavaScript-like syntax of Solidity, including the use of camelCasing, semicolons, and inline comments. A few notable differences exist as well: the type system, a compiler version declaration, and new keywords. Hopefully this example was simple enough to convey the concepts, but the intricacies of the language are beyond the scope of this article. You'll find links to mor ...
```
oooooh.....competition :D

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-3.mdx:
```diff
@@ -0,0 +1,209 @@
+
+## How does a contract get on the blockchain?
+
+Earlier in this blog series, you may recall reading that the only way to make changes to the state of the Ethereum blockchain is via transactions. That remains true for deploying new contracts.
```
`blog` can we simply omit it?

elPiablo replied:
I would ask `..., do you recall ...?` instead of the statement `..., you may recall`. It gives just a little jolt of ownership to the readers (it personalises).

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-3.mdx:
```diff
@@ -0,0 +1,209 @@
+Tools like web3.py offer slightly more human-friendly ways to go about this. One of the other outputs of a compilation is the contract's **ABI**, and some additional metadata.
+
+<Callout emoji="💡" size="md" variant="info">
+**Note**: ABI stands for **Application Binary Interface**. An ABI is a machine-readable data blob that conveys how a contract can be interacted with – which functions are available and expected data types. It's some JSON that you pass into an Ethereum library (e.g., web3.py, ethers.js, etc.), so that it can provide you with a human-friendly interface. Does the name make sense now? An ABI communic ...
```
Love the question!

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-3.mdx:
```diff
@@ -0,0 +1,209 @@
+Over the years, standards for various digital assets have proposed, debated, and agreed upon, providing some foundational building blocks for more complex contracts. Among the most notable are the ERC-20 token standard (i.e., fungible tokens) and the ERC-721 standard (i.e., non-fungible tokens or "NFTs").
+
+<Callout emoji="💡" size="md" variant="info">
+**Note**: To save you the Google search, _fungible_ means interchangeable or indistinguishable. In other words, if you own 100 fungible tokens, it doesn't make any difference which 100 tokens they are. NFTs, on the other hand, may each have unique qualities, so the particular token you own is significant.
```
`Google` search? you surprise me, Mr Garreau :D

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-3.mdx:
```diff
@@ -0,0 +1,209 @@
+
+Once deployed, the `MyToken` contract has access to all the functions defined in OpenZeppelin's `ERC20` contract implementation. Conveniently, you don't have to reinvent the wheel and can focus on what makes your token contract unique.
+
+Beyond inheritance, contracts have the ability to interact with other deployed contracts or even serve as a _factory_ or _proxy_ for deploying still more contracts! Those concepts are good topics for future blog posts.
```
`blog posts`

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-3.mdx:
```diff
@@ -0,0 +1,209 @@
+
+While the blockchain is said to be immutable, there are patterns of writing smart contracts that can support upgradeability. These patterns introduce trade-offs which may or may not make sense for your use case. The specifics are beyond the scope of this introductory lesson.
+
+## Starting off on your own
```
Generally speaking, there are two sorts of links here. Help resources, and learning resources. 

Before we started Academy, we tasked ourselves with creating two Notion tracks for  web2 and web3, and two Discord channels to manage them proactively. It was similar sorts of rescources we were compiling. We eventually abandoned it, and started building School of Code => Academy. 

I don't think anyone wants to deny advice to a learner to good quality content mentioned here, but at the same time, `we're abandoning the mission and vision of this education platform`, and in my mind, what our DAO has potentially to offer.

Culture building is really important. Raza, or DappaDan, or any other potential authors might just look in at this, and think `oh links are okay, or an easy way out` of writing their own self-supporting material. And then our platform is saturated with links, and when we do get our own Ethereum Stack Exchange (e.g. Peerahna) up and running, moderators are going to be tasked with trying to answer questions far removed from their/our own work in the beginning till it becomes self-supporting. We don't have the bandwidth for that. 

Starting humbly can also be an attractive aspect for budding authors. They'll know that the attention-factor on Academy will be all theirs.

I'd say a better approach would be to leave in a couple of help links (e.g. Solidity by Example and Remix, or thereabouts), and pursue a more homegrown approach, suggesting an invite inwards: https://handbook.developerdao.com/

wolovim replied:
i altogether nuked this section and links within, encouraging continuation within Academy. agree on the importance of simple focus at this stage in the learners journey.

elPiablo commented:
Hey @wolovim 
All a really nice set of tutorials.
Although I'm wondering what you see in the future for these three tutorials with respect to these structures: https://www.notion.so/developerdao/Guidelines-for-Educational-Content-5e702be7dbb74bceb2b06617019d3f8d for many different reasons/actors

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-3.mdx:
```diff
@@ -0,0 +1,209 @@
+
+In reality, most users aren't reading the source code of each smart contract they interact with, but most wisely won't touch a contract if that source code isn't verified (e.g., on sourcify or Etherscan) and vetted (e.g., audited) by industry veterans.
+
+Consider the alternative: if contracts are black boxes, there's nothing to stop a bad actor from publishing a contract that appears innocuous, but actually grants themselves the ability to move your assets. Today, bad actors can deploy such a contract and try to lure users in via social engineering, but often wallet interfaces will warn users when code is unverified and to proceed with caution.
```
`harmless` might save folks like me having to search in a dictionary for `innocuous` :D

elPiablo commented on src/pages/lessons/fundamentals/eth-intro-part-3.mdx:
```diff
@@ -0,0 +1,209 @@
+
+The Ethereum blockchain has a great deal of value flowing through it. In previous articles, we discussed the flow of **ether** from one user to another via transactions. However, the network is capable of more sophisticated interactions, and those use cases are enabled by smart contracts.
+
+Let's start with a simple definition: a **smart contract** is code (i.e., a computer program) deployed to the blockchain. As for the buzzword name, _contract_ conveys the relative permanence of these programs as they determine how assets can change hands and _smart_ is a nod to their programmability. For brevity, they're commonly referred to as just contracts, and we'll follow suit.
```
This is some pretty high level English: `For 'brevity', they're commonly referred to as just contracts, and we'll 'follow suit'.` could you simplify just a part of it?

elPiablo reviewed (commented):
Hey @wolovim
All a really nice set of tutorials. Really well written.
Although I'm wondering what you see in the future for these three tutorials with respect to these structures: https://www.notion.so/developerdao/Guidelines-for-Educational-Content-5e702be7dbb74bceb2b06617019d3f8d for many different reasons/actors

wolovim committed:
chore: address pr review on pt 2

wolovim committed:
chore: address pr review on pt 3

wolovim committed:
chore: remove quiz answer

wolovim commented:
thanks for thorough review @elPiablo 🖖 

ive addressed all feedback with exception of structural updates to the intros. i'll suggest this would be wise to wait until porting over to v2 to optimize for.

elPiablo reviewed (approved):
An absolutely brilliant three part track of craftsmanship. You're a genius and an artist at that! So so happy to see this finally going live. So happy, you wouldn't believe. 
I added two new comments in particular in reply to previous comments in there (they are all now marked as resolved, so you may need to search for them). Nothing to hold up the PR, but we need to return to them as a matter of process, and not as nitpicks.
So LFG!!! Merge this to the Academy!!

wolovim commented:
thank you @elPiablo - happy to keep iterating on your few unresolved points in another PR. dont let me forget about em in v2 ;)

elPiablo commented:
Hey @wolovim 
Don't forget to merge into `main` from `staging` branch :)
